### PR TITLE
Userland: Start to display some metadata from images

### PR DESCRIPTION
**Pull request by LucasChollet**
This PR is a mix from final changes to get the TIFF decoder to expose some valuable metadata, the addition of an API to ImageDecoderPlugin (that needs the more thoroughly review IMO), and some finally some GUI addition to display the metadata in `FileManager`.

image

### CODE REVIEW

**Comment by warpdesign** [comment]
👍
Just wondering, IFF may have various chunks containing text information describing the files: author, copyright, artist, name, etc..

How would these fit here? Shouldn't we have some kind of generic tag so that we could fill it in with generic information?

File: Userland/Libraries/LibGfx/ImageFormats/ImageDecoder.h [hunk 1]
```diff
@@ -28,6 +30,15 @@ struct VectorImageFrameDescriptor {
     int duration { 0 };
 };
 
+struct BasicMetadata {
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/ImageDecoder.h:33]: wdyt about making this a pure virtual interface and having some class that derives from it and that contains a ref to ExifMetadata internally (BasicExifMetadata or what)? Other image formats could then also implement this interface.

How did you come up with manufacturer, model, software, creation_time, author as the entries here?

The concrete metadata formats that matter are probably exif, iptc, xmp.

PNG has this https://www.w3.org/TR/2003/REC-PNG-20031110/#11keywords

TIFF probably has something like that list somewhere too.

**Review by nico — COMMENTED** [review]
Seems alright-ish to me. It sidesteps most of the interesting questions about metadata, which is probably the thing to do if you want to get something in sooner :P

The interesting questions about metadata are imho:

1. What to do with metadata that needs a lot of memory, like thumbnails, makernotes, etc

2. How to handle metadata entries where multiple tags produce a single user-visible string (e.g. gps information, where one tag has e.g. latitude number and another the latitude sign (having a "MetadataIterator" thingy that yields key/value string pairs and implementing that on top of ExifMetadata to handle cases like this is probably the way to go)

3. What to do with things that are technically metadata but actually required to display images correctly (mostly the Exif Orientation tag)

**Comment by LucasChollet** [comment]
> It sidesteps most of the interesting questions about metadata

Yep, let's do the simple stuff now and figure out the complex part later on :smile:

My two cents on your questions:

1. I never thought of makernotes, but thumbnails should be a separate API, they are used in a very different context IMO.
2. A single entry in `main_tags` (if they have their place there) and multiple when iterating over the HashMap in ExifMetadata.
3. As we started to do in the TIFF decoder, this should be available at decoding time, so the decoder should be able to access it. But after that, the Orientation tag is not special.

File: Userland/Libraries/LibGfx/ImageFormats/ImageDecoder.h (hunk 1 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ImageDecoder.h]: > How did you come up with manufacturer, model, software, creation_time, author as the entries here?

I tried to have something fairly compatible with most format, but that's probably not doable. So I tried something else: I completely changed the design to be way more flexible. This is done by having the basic metadata already being passed as an iterable. So every format can provide their full tag list and the important subset for the property window.

File: Userland/Libraries/LibGfx/TIFFGenerator.py [hunk 2]
```diff
@@ -124,6 +126,9 @@ class ExtraSample(EnumWithExportName):
     Tag('285', [TIFFType.ASCII], [], None, "PageName"),
     Tag('296', [TIFFType.UnsignedShort], [1], ResolutionUnit.Inch, "ResolutionUnit", ResolutionUnit),
     Tag('339', [TIFFType.UnsignedShort], [], SampleFormat.Unsigned, "SampleFormat", SampleFormat),
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGfx/TIFFGenerator.py:128]: Not a part of this PR, but this one is out of order.

**Review by AtkinsSJ — APPROVED** [review]
Looks good to me, and very nice to get this data exposed, but there's a conflict.

File: Userland/Libraries/LibGfx/TIFFGenerator.py (hunk 2 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/TIFFGenerator.py:128]: :redlinus:

**Pull request merged**